dashboard/app/controllers/lti_v1_controller.rb

```diff
@@ -109,6 +109,18 @@ def authenticate
       return log_unauthorized(exception)
     end
 
+    # Schoology has multiple contexts that will launch LTI tools in an iframe.
```

Comment:
Initially I had this iframe check run for all LTI launches, at the top of the `authenticate` controller action. However, this has a couple downsides.

1. It redirects the request _before_ doing any jwt validation and authentication checks
2. It becomes harder to unit test this controller action, as the first thing it does is redirect to another page, unless the `new_tab` param is included in the request

I indexed on doing all the jwt validation and auth checks before redirecting, as I would rather deny the call before redirecting. The downside of this is now the redirect happens after a db call to retrieve the LTI Integration, which means this db request will happen two times now per LTI launch for Schoology.

I'm open to discussing further options here. In theory, we should be able to determine if the LMS is opening us in an iframe by the `https://purl.imsglobal.org/spec/lti/claim/launch_presentation`, however in every id_token I've examined from both Canvas and Schoology, no matter the placement, this value is always `iframe`.

Reply:
Could we cache integrations for use in the auth routes? Would save us some lookups if all we need to do is check that it exists and lookup the various URLs. Might not be needed right away, but maybe down the road.

Reply:
Oh interesting. Mario had suggested the idea of caching on the first "round" then reading from the cache when the iframe flow redirects back here. I like the idea of caching integrations as a whole, to be used by multiple LTI launch requests though. We already have the write_cache and read_cache helper functions. It may be an over optimization right now, but it might be the right thing to do.